pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -188,3 +351,55 @@ func getInfraMachineTemplateFromProvider(platform configv1.PlatformType) (client
 		return nil, fmt.Errorf("%w: %s", errPlatformNotSupported, platform)
 	}
 }
+
+// updateMAPIMachineSetSpecFromCAPI updates the MAPI MachineSet spec based on the CAPI MachineSet.
+func updateMAPIMachineSetSpecFromCAPI(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet, capiMachineSet *capiv1beta1.MachineSet, newProviderSpec machinev1beta1.ProviderSpec) {
+	logger := log.FromContext(ctx).WithValues("function", "updateMAPIMachineSetSpecFromCAPI")
+
+	// Update the ProviderSpec
+	mapiMachineSet.Spec.Template.Spec.ProviderSpec = newProviderSpec
+
+	// Update labels (CAPI labels take priority)
+	mapiMachineSet.Spec.Template.Labels = util.MergeMaps(mapiMachineSet.Spec.Template.Labels, capiMachineSet.Spec.Template.Labels)
+
+	// Update taints ?? I can't find 'em on the CAPI MachineSet
```

Comment:
I'm not entirely sure how to handle taints here. 

IIUC (from grepping around the capi codebase) this may be taken care of by a bootstrap provider?